rcl/src/rcl/node.c

```diff
@@ -542,6 +564,173 @@ rcl_get_disable_loaned_message(bool * disable_loaned_message)
+    }
+
+    if (!response.successful) {
+      rosidl_runtime_c__String__assign(
+        &response.failure_reason,
+        "Failed to populate response");
+    }
+  } else {
+    response.successful = false;
+    RCUTILS_LOG_INFO_NAMED(
+      ROS_PACKAGE_NAME,
+      "Type '%s' not found in type cache",
+      request.type_hash.data);
+    rosidl_runtime_c__String__assign(
+      &response.failure_reason,
+      "Type not currently used by this node");
+  }
+
+  if (RCL_RET_OK != rcl_send_response(
+      &node->impl->get_type_description_service,
+      &request_header, &response))
+  {
+    RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Failed to send response");
+  }
+
+cleanup:
+  type_description_interfaces__srv__GetTypeDescription_Request__fini(&request);
+  type_description_interfaces__srv__GetTypeDescription_Response__fini(
+    &response);
+}
+
+rcl_ret_t rcl_node_type_description_service_init(rcl_node_t * node)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(node->impl, RCL_RET_NODE_INVALID);
+
+  if (rcl_service_is_valid(&node->impl->get_type_description_service)) {
+    return RCL_RET_ALREADY_INIT;
+  }
+  rcl_reset_error();  // Reset the error message set by rcl_service_is_valid()
+
+  char * serviceName = NULL;
+  const rosidl_service_type_support_t * type_support =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(
+    type_description_interfaces, srv,
+    GetTypeDescription);
+  rcl_service_options_t service_ops = rcl_service_get_default_options();
+  rcl_allocator_t allocator = node->context->impl->allocator;
+
+  // Construct service name
+  rcl_ret_t ret = rcl_node_resolve_name(
+    node, "~/get_type_description",
+    allocator, true, true, &serviceName);
+  if (RCL_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(
+      "Failed to construct ~/get_type_description service name");
+    goto cleanup;
+  }
+
+  // Initialize service
+  ret = rcl_service_init(
+    &node->impl->get_type_description_service, node,
+    type_support, serviceName, &service_ops);
+  if (RCL_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(
+      "Failed to initialize ~/get_type_description service");
+    goto cleanup;
+  }
+
+  // Set service callback
+  ret = rcl_service_set_on_new_request_callback(
+    &node->impl->get_type_description_service,
+    rcl_node_get_type_description_service_callback, node);
+  if (RCL_RET_OK != ret) {
+    rcutils_reset_error();
+    if (RCL_RET_OK !=
+      rcl_service_fini(&node->impl->get_type_description_service, node))
+    {
+      RCUTILS_LOG_ERROR_NAMED(
+        ROS_PACKAGE_NAME,
+        "Failed to deinitialize ~/get_type_description service.");
+    }
+    node->impl->get_type_description_service = rcl_get_zero_initialized_service();
+    rcutils_reset_error();
+    RCL_SET_ERROR_MSG(
+      "Failed to set callback for ~/get_type_description service");
+  }
+
+cleanup:
+  if (serviceName) {
+    allocator.deallocate(serviceName, allocator.state);
+  }
+
+  return ret;
```

Comment:
I think you can revamp the error handling here so that this is simpler.  In particular:
```suggestion
  ret = rcl_service_init(
    &node->impl->get_type_description_service, node,
    type_support, serviceName, &service_ops);
  allocator.deallocate(serviceName, allocator.state);
  if (RCL_RET_OK != ret) {
    return ret;
  }

  // Set service callback
  ret = rcl_service_set_on_new_request_callback(
    &node->impl->get_type_description_service,
    rcl_node_get_type_description_service_callback, node);
  if (RCL_RET_OK != ret) {
    goto fail_fini_service;
  }

  return RCL_RET_OK;

fail_fini_service:
  if (RCL_RET_OK !=
    rcl_service_fini(&node->impl->get_type_description_service, node))
  {
    RCUTILS_LOG_ERROR_NAMED(
      ROS_PACKAGE_NAME,
      "Failed to deinitialize ~/get_type_description service.");
  }

  return ret;
```